### ISSUE TITLE
migration:create changes non-nullable to nullable, changes it back and back and forth for enum and custom type properties with default value

### ISSUE THREAD

**Issue by ddadaal** [mikro-orm/mikro-orm#1559]
**Describe the bug**
Using mysql driver, for `enum` properties with default value and custom type properties with default values, repeated `migration:create` runs creates new migration files which removes the properties' `not null`, adds `not null`, removes it, ands it back and back and forth.

**To Reproduce**
A repo with clearer description and steps to reproduce: https://github.com/ddadaal/mikro-orm-migration-changing-nonnullable-to-nullable

**Expected behavior**
No new migration files should be generated after the first `migration:create`, since there is no change at all in the entity.

**Versions**

| Dependency   | Version  |
| ------------ | -------- |
| node         | v14.15.1 |
| typescript   | 4.2.3    |
| mikro-orm    | 4.4.4    |
| mysql-driver | 4.4.4.   |

**Comment by B4nan** [comment]
Afaik in your custom type, `"DECIMAL(19, 4)"` should be `"DECIMAL(19,4)"` (no space) - or maybe it should be `numeric`, not sure right now, but you need to use the actual type name as the DB stores it, one of those is just an alias. 

Also, this has nothing to do with migrations, this is schema generator/diffing issue, migrations are just a layer on top of it.

**Comment by ddadaal** [comment]
Thanks for the advice, but actually in my production code, `DECIMAL(19, 4)` works fine, 

<img width="687" alt="screenshot-decimal" src="https://user-images.githubusercontent.com/8363856/111033301-7e38d700-844b-11eb-9a68-a176109ebe4c.png">

and I am mapping it into a [bignumber.js](https://mikemcl.github.io/bignumber.js/) instance like so:

```ts
import { Type } from "@mikro-orm/core";
import { BigNumber as Decimal } from "bignumber.js";

const precision = 4;

const moneyColumnType = `DECIMAL(19, ${precision})`;

Decimal.set({ DECIMAL_PLACES: precision });

export { Decimal };

export class DecimalType extends Type<Decimal | undefined, string | undefined> {
  convertToDatabaseValue(value: Decimal | string | undefined): string | undefined {
    if (!value) { return value;}
    return value.toString();
  }

  convertToJSValue(value: string | undefined): Decimal | undefined {
    if (!value) { return undefined; }
    return new Decimal(value);
  }

  getColumnType(): string {
    return moneyColumnType;
  }
}
```

And it seems to work fine as well. 

So should I report the issue to umzug?

**Comment by B4nan** [comment]
> Thanks for the advice, but actually in my production code, DECIMAL(19, 4) works fine,

Have you tried it? This is not about "it works for creating the column", this is about diffing the column type, and for that the type needs to be normalized to what the information schema stores. And there I believe it has to be without the space - someone already reported this very same issue and without the space it was working as expected. 

Umzug has nothing to do with this, neither have the migrations - as I mentioned, this is about schema diffing inside `SchemaGenerator`.

**Comment by ddadaal** [comment]
I added some tests in the reproduce repo to test the mapping and updating of the `bignumber.js` instance and `DECIMAL(19, 4)` db column, and I think it to some extent proves that the mapping should work as expected.

And I am sorry I got confused back then and kept thinking it was about migration. I fully understand it's about `SchemaGenerator` now.

**Comment by ddadaal** [comment]
OK I found a workaround for this issue:

1. Remove the space in `DECIMAL(19, 4)` to `DECIMAL(19,4)`
2. Use `defaultRaw` instead of `default` for enum and custom type default values

After these modifications, the diffing generates correct result.

Here are the code lines that I found related to this issue:

https://github.com/mikro-orm/mikro-orm/blob/288eee4b9bc92d60bca7e2cec8c32c04ed322714/packages/knex/src/schema/SchemaHelper.ts#L196-L197

For the specific custom type in my case, the regex matches `(19,4)` but not `(19, 4)` (with space). As a result, `DECIMAL(19, 4)` is changed to `decimal` which equals to the db type `decimal`, but `DECIMAL(19, 4)` is only changed `decimal(19, 4)` which `!== "decimal"`.  This seems to be a easy fix, or maybe it's the desired behavior and I should not add the space in the first place?

https://github.com/mikro-orm/mikro-orm/blob/288eee4b9bc92d60bca7e2cec8c32c04ed322714/packages/knex/src/schema/SchemaHelper.ts#L224-L231

For the enum with default value, the diffing compares the `defaultValue` of db column with `defaultRaw` of property. The column prop from db has `defaultValue` of `'A'`, but the `defaultRaw` of property is `'"A"'`, which are not equal.

```
{
  name: 'a_enum',
  type: 'enum',
  ...
  defaultValue: 'A',
  ...
} {
  name: 'aEnum',
  ...
  defaultRaw: "'A'"
}
```

But I think the root cause should be inside this function, or `knex` which actually changes the column:

https://github.com/mikro-orm/mikro-orm/blob/288eee4b9bc92d60bca7e2cec8c32c04ed322714/packages/knex/src/schema/SchemaGenerator.ts#L435-L455

For the enum property, when the workaround is not applied, the diffing thinks only the default value is different, so only 

https://github.com/mikro-orm/mikro-orm/blob/288eee4b9bc92d60bca7e2cec8c32c04ed322714/packages/knex/src/schema/SchemaGenerator.ts#L451

is executed, which unexpectedly removes `non null` in the final result.
